### PR TITLE
Adding fp16 support

### PR DESCRIPTION
**Pull request by rtabbara**

### CODE REVIEW

File: src/cuda_eval.cpp [hunk 1]
```diff
@@ -194,7 +197,7 @@ void jitc_cuda_assemble(ThreadState *ts, ScheduledGroup group,
                 fmt("    mad.wide.u32 %rd0, %r0, $a, %rd0;\n", v);
 
             if (vt != VarType::Bool) {
-                fmt("    $s$t $v, [%rd0];\n",
```
Comment (wjakob) [inline src/cuda_eval.cpp:197]: Would the existing `$b` not address the problem in a more general way here?

File: include/drjit-core/half.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
+/*
+    drjit-core/half.h -- minimal half precision number type
+
+    Copyright (c) 2021 Wenzel Jakob <wenzel.jakob@epfl.ch>
```
Comment (wjakob) [inline include/drjit-core/half.h:4]: Hm, it's weird to list me as copyright author given that you wrote this code. Why don't you remove it for now or put your name? Or we can change the copyright note to list the lab in a later commit.

File: src/cuda_eval.cpp [hunk 3]
```diff
@@ -214,7 +217,7 @@ void jitc_cuda_assemble(ThreadState *ts, ScheduledGroup group,
                 params_type, params_base, v, v);
 
             if (vt != VarType::Bool) {
-                fmt("    st.global.cs.$t [%rd0], $v;\n", v, v);
+                fmt("    st.global.cs.$T [%rd0], $v;\n", v, v);
```
Comment (wjakob) [inline src/cuda_eval.cpp:220]: ditto

File: src/cuda_eval.cpp [hunk 4]
```diff
@@ -336,7 +339,7 @@ void jitc_cuda_assemble_func(const char *name, uint32_t inst_id,
 
         if (v->vcall_iface) {
             if (vt != VarType::Bool) {
-                fmt("    ld.param.$t $v, [params+$o];\n", v, v, v);
+                fmt("    ld.param.$T $v, [params+$o];\n", v, v, v);
```
Comment (wjakob) [inline src/cuda_eval.cpp:342]: ditto

File: src/cuda_eval.cpp [hunk 5]
```diff
@@ -359,7 +362,7 @@ void jitc_cuda_assemble_func(const char *name, uint32_t inst_id,
                     "is happening now). This is not allowed.", sv.index);
 
             if (vt != VarType::Bool)
-                fmt("    ld.global.$t $v, [data+$u];\n",
+                fmt("    ld.global.$T $v, [data+$u];\n",
```
Comment (wjakob) [inline src/cuda_eval.cpp:365]: ditto

File: src/cuda_eval.cpp [hunk 6]
```diff
@@ -383,7 +386,7 @@ void jitc_cuda_assemble_func(const char *name, uint32_t inst_id,
         uint32_t vti = v->type;
 
         if ((VarType) vti != VarType::Bool) {
-            fmt("    st.param.$t [result+$u], $v;\n", v, offset, v);
+            fmt("    st.param.$T [result+$u], $v;\n", v, offset, v);
```
Comment (wjakob) [inline src/cuda_eval.cpp:389]: ditto

File: src/cuda_eval.cpp [hunk 7]
```diff
@@ -430,8 +433,17 @@ static void jitc_cuda_render_var(uint32_t index, const Variable *v) {
             break;
 
         case VarKind::Sqrt:
-            fmt(jitc_is_single(v) ? "    sqrt.approx.ftz.$t $v, $v;\n"
-                                  : "    sqrt.rn.$t $v, $v;\n", v, v, a0);
+            if (jitc_is_half(v))
+                fmt("    cvt.f32.f16 %f0, $v;\n"
```
Comment (wjakob) [inline src/cuda_eval.cpp:437]: Given that this pattern repeats in a few places, how about we have a special per-instruction flag stating whether such a conversion is needed? Then those two lines could be added before/after the switch statement. And we can temporarily overwrite the `v->reg_index` so that `$v` reaches the right operand.

 We'll need to reuse this approach if we ever add 8-bit arithmetic, etc.

**Review by wjakob — COMMENTED** [review]
Hi Rami,

this already looks fantastic and is nice and short. A few first quick comments from me:

File: tests/loop.cpp [hunk 8]
```diff
@@ -11,13 +11,13 @@ TEST_BOTH(01_record_loop) {
         for (uint32_t j = 0; j < 2; ++j) {
             UInt32 x = arange<UInt32>(10);
             Float y = zero<Float>(1);
-            Float z = 1;
+            Float z = Float(1);
```
Comment (wjakob) [inline tests/loop.cpp:14]: I am curious: why does this not work? Having an implicit conversion from an integer to a float is probably a good idea.

File: src/op.cpp [hunk 9]
```diff
@@ -291,10 +295,10 @@ uint32_t jitc_var_shift(const VarInfo &info, uint32_t index, uint64_t amount) {
 
 // --------------------------------------------------------------------------
 
-template <typename T, enable_if_t<!std::is_signed_v<T>> = 0>
+template <typename T, enable_if_t<!std::is_signed<T>::value> = 0>
```
Comment (wjakob) [inline src/op.cpp:298]: I am curious: why was this change needed? The two should be the same, and the former version was shorter.

File: include/drjit-core/half.h [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+template<> struct is_signed<drjit::dr_half> : true_type { };
+NAMESPACE_END(std)
+
+
+NAMESPACE_BEGIN(drjit)
+struct dr_half {
+
+    uint16_t value;
+
+    #define DRJIT_IF_SCALAR template <typename Value, enable_if_t<std::is_arithmetic<Value>::value> = 0>
```
Comment (njroussel) [inline include/drjit-core/half.h:41]:
```suggestion
    #define DRJIT_IF_SCALAR template <typename Value, enable_if_t<!std::is_arithmetic_v<Value>>
```

File: include/drjit-core/half.h [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+    dr_half() = default;
+
+    DRJIT_IF_SCALAR dr_half(Value val) : value(float32_to_float16(float(val))) { }
```
Comment (njroussel) [inline include/drjit-core/half.h:45]: Using `float()` as a constructor rather than a C-style cast is a bit weird here. No?

File: src/cuda_eval.cpp [hunk 12]
```diff
@@ -1271,16 +1308,16 @@ void jitc_var_vcall_assemble_cuda(VCall *vcall, uint32_t vcall_reg,
         const Variable *v2 = &it->second;
         uint32_t size = type_size[v2->type];
 
-        const char *tname = type_name_ptx[v2->type],
-                   *prefix = type_prefix[v2->type];
+        const char *tname               = type_name_ptx_bin[v2->type],
+                   *prefix              = type_prefix[v2->type];
```
Comment (njroussel) [inline src/cuda_eval.cpp:1312]: No need to align these

File: src/llvm_eval.cpp [hunk 13]
```diff
@@ -733,10 +733,22 @@ static void jitc_llvm_render_var(uint32_t index, Variable *v) {
                                     : "    $v = fptosi $V to $T\n",
                     v, a0, v);
             } else if (jitc_is_float(v) && jitc_is_float(a0)) {
-                fmt(type_size[v->type] > type_size[a0->type]
-                        ? "    $v = fpext $V to $T\n"
-                        : "    $v = fptrunc $V to $T\n",
-                    v, a0, v);
+                // On x86, direct double-half casting relies on external builtin function call
+                // unless AVX512_FP16 instructions are supported so split casting into two-steps
+                // i.e. double<->float<->half
```
Comment (njroussel) [inline src/llvm_eval.cpp:738]: :+1:

File: src/op.cpp [hunk 14]
```diff
@@ -291,10 +295,10 @@ uint32_t jitc_var_shift(const VarInfo &info, uint32_t index, uint64_t amount) {
...
 T eval_neg(T v) { return T(-(std::make_signed_t<T>) v); }
 
-template <typename T, enable_if_t<std::is_signed_v<T>> = 0>
+template <typename T, enable_if_t<std::is_signed<T>::value> = 0>
```
[the remark this answers is not in the document]
Comment (njroussel) [reply src/op.cpp:301]: Nevermind, there are quite a few more on the following lines

File: src/op.cpp [hunk 15]
```diff
@@ -341,7 +345,7 @@ uint32_t jitc_var_not(uint32_t a0) {
 // --------------------------------------------------------------------------
 
 template <typename T, enable_if_t<std::is_floating_point_v<T>> = 0>
-T eval_sqrt(T value) { return std::sqrt(value); }
+T eval_sqrt(T value) { using std::sqrt, drjit::dr_half; return ::sqrt(value); }
```
Comment (njroussel) [inline src/op.cpp:348]: Very clean!

File: src/var.cpp [hunk 16]
```diff
@@ -410,6 +410,7 @@ void jitc_value_print(const Variable *v, bool graphviz = false) {
         break;
 
     switch ((VarType) v->type) {
+        case VarType::Float16: JIT_LITERAL_PRINT(uint16_t, unsigned, "%u");
```
Comment (njroussel) [inline src/var.cpp:413]: `jitc_value_print` is meant for human-readable outputs. I think it would make sense if printed the floating point value rather than its `uint16_t` representation.

Comment (njroussel) [reply src/var.cpp:413]: I see now that you're doing this in `jitc_var_str()`. I think we would something similar here.

File: include/drjit-core/half.h (hunk 11 above)
Comment (njroussel) [reply include/drjit-core/half.h:45]: We should also maybe explicitly instantiate this constructor for all arithmetic types. Otherwise we won't get any implicit constructor calls like `dr_half my_var = 3u;` in other libraries.

**Review by njroussel — COMMENTED** [review]
Looking good! I added a few suggestions for improvements :+1:

**Comment by wjakob** [comment]
How about we rename the half type from `dr_half` to just `half`?

**Comment by wjakob** [comment]
A stupid question: out of all the various arithmetic operations, why is sqrt treated specially?

File: src/cuda_eval.cpp (hunk 1 above)
Comment (rtabbara) [reply src/cuda_eval.cpp:197]: Yes, you're right..I've fixed that up.

File: src/cuda_eval.cpp (hunk 3 above)
Comment (rtabbara) [reply src/cuda_eval.cpp]: I was trying to be a bit too clever by adding a special type character just to handle the fact PTX instructions for a lot of the data move/load/store operations can't be specified by a `fp16`  type and instead need `b16` (keeping all the other types the same). But I agree, just having `$b` for everything doesn't really hurt readability and reduces bloat so I've made that switch.

File: src/cuda_eval.cpp (hunk 7 above)
Comment (rtabbara) [reply src/cuda_eval.cpp]: Makes sense, that change has been added!

File: tests/loop.cpp (hunk 8 above)
Comment (rtabbara) [reply tests/loop.cpp]: This has been fixed

File: src/op.cpp (hunk 9 above)
Comment (rtabbara) [reply src/op.cpp]: Specializing `std::is_signed`, `std::is_arithmetic` etc. is officially considered undefined behaviour, which we were using for the `half` type. For GCC `is_signed_v` worked fine, but then I was having issues compiling with Clang which is why I made this change. But then I realised MSVC was also broken and in the end I thought it was safest to define our own `drjit::is_signed` etc. traits to avoid relying on undefined behaviour. So hope that's ok.

File: include/drjit-core/half.h [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+        #elif defined(__aarch64__)
+            return (float)memcpy_cast<__fp16>(value);
+        #else
+            jit_fail("Unsupported architecture");
+        #endif
+    }
+
+    static dr_half sqrt(dr_half h) { 
+        #if defined(__aarch64__)
+            return (float)vsqrth_f16(memcpy_cast<__fp16>(h.value));
```
[the remark this answers is not in the document]
Comment (rtabbara) [reply include/drjit-core/half.h]: That's a good point. Fixed!

File: src/op.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (rtabbara) [reply src/op.cpp]: As mentioned before, this was my hacky way to get around undefined behaviour of specialising these type traits

File: include/drjit-core/half.h (hunk 11 above)
Comment (rtabbara) [reply include/drjit-core/half.h]: Fixed!

File: include/drjit-core/half.h [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+#include <drjit-core/jit.h>
+#include <drjit-core/traits.h>
+
+#include <cmath>
+#include <cstring>
+
+#if defined __aarch64__
+#include <arm_fp16.h>
+#elif defined(__F16C__)
+#include <emmintrin.h>
```
Comment (wjakob) [inline include/drjit-core/half.h:19]: I think you can remove `<emmintrin.h>` here. AFAIK `<immintrin.h>` includes everything.

File: src/cuda_eval.cpp [hunk 19]
```diff
@@ -408,6 +408,26 @@ static void jitc_cuda_render_var(uint32_t index, const Variable *v) {
              *a2 = v->dep[2] ? jitc_var(v->dep[2]) : nullptr,
              *a3 = v->dep[3] ? jitc_var(v->dep[3]) : nullptr;
 
+    bool f32_upcast = jitc_is_half(v) && var_kind_fp16_unsupported_cuda[v->kind];
+
+    if (f32_upcast) {
+        Variable* b = const_cast<Variable*>(v);
+        b->type = (uint32_t)VarType::Float32;
+        for (size_t i = 0; i < 4; ++i) {
+            Variable* dep = b->dep[i] ? jitc_var(b->dep[i]) : nullptr;
```
Comment (wjakob) [inline src/cuda_eval.cpp:417]: (Despite the comment below). Would it not be easier to understand to merge these two loops? You can add a check to see if the variable has already been set to `Float32`. I think the current implementation may generate the `cvt.f32.f16` instruction multiple times for operations that receive the same operand twice.

File: src/cuda_eval.cpp [hunk 20]
```diff
@@ -802,6 +842,8 @@ static void jitc_cuda_render_scatter(const Variable *v,
     const char *op = reduce_op_name[v->literal];
     const ThreadState *ts = thread_state_cuda;
 
+    if (jitc_is_half(value) && v->literal != 0)
```
Comment (wjakob) [inline src/cuda_eval.cpp:845]: This is not a nice way to catch this special case because it will crash the Python session. Can we already do this in `op.cpp` where the associated scatter variable would be created?

File: src/cuda_eval.cpp [hunk 21]
```diff
@@ -920,15 +964,16 @@ static void jitc_cuda_render_scatter_kahan(const Variable *v, uint32_t v_index)
...
+    const char* red_atom_suffix = jitc_is_half(value) ? ".noftz" : "";
 
     fmt("    {\n"
         "        .reg.$t %before, %after, %value, %case_1, %case_2;\n"
         "        .reg.$t %abs_before, %abs_value, %result;\n"
         "        .reg.pred %cond;\n"
         "\n"
-        "        mov.$t %value, $v;\n"
-        "        atom.global.add.$t %before, [%rd2], %value;\n"
+        "        mov.$b %value, $v;\n"
```
Comment (wjakob) [inline src/cuda_eval.cpp:975]: If float16 atomics aren't properly supported, then I don't understand why there are changes here. I would prefer to keep the patch small in this case, unless I am misunderstanding something?

File: src/var.cpp [hunk 22]
```diff
@@ -172,6 +172,94 @@ const char *var_kind_name[(int) VarKind::Count] {
     "extract"
 };
 
+const bool var_kind_fp16_unsupported_cuda[(int) VarKind::Count] {
```
Comment (wjakob) [inline src/var.cpp:175]: would it not be easier to understand to encode the inverse? (i.e., `fp16_supported`)

**Review by wjakob — COMMENTED** [review]
Hi Rami,

Thanks for making those changes — here are a few more comments from me.

-Wenzel

File: src/op.cpp (hunk 9 above)
Comment (wjakob) [reply src/op.cpp]: Making our own specialized version of these classes seems fine and safest.
